From ffb1db21675b49383272e15b7580c6d8177b4e06 Mon Sep 17 00:00:00 2001
From: wanlong <long.wan@spacemit.com>
Date: Wed, 23 Oct 2024 16:44:35 +0800
Subject: [PATCH 1194/1448] wireless: rtl8852bs: fix wiphy lockdep warning

WARNING: CPU: 7 PID: 98 at net/wireless/reg.c:3144 reg_process_self_managed_hint+0x154/0x1aa
Modules linked in: 8021q garp stp mrp llc 8852bs
CPU: 7 PID: 98 Comm: kworker/7:1 Not tainted 6.6.36 #23
Hardware name: spacemit k1-x deb1 board (DT)
Workqueue: events_freezable mmc_rescan
epc : reg_process_self_managed_hint+0x154/0x1aa
 ra : reg_process_self_managed_hint+0x150/0x1aa
epc : ffffffff80eecbd8 ra : ffffffff80eecbd4 sp : ffffffc8009b3730
 gp : ffffffff83541fe0 tp : ffffffd903a23380 t0 : ffffffd90511ce00
 t1 : 0000000000000001 t2 : 0000000000000000 s0 : ffffffc8009b37c0
 s1 : 0000000000000000 a0 : 0000000000000000 a1 : ffffffd903976810
 a2 : 0000000000000000 a3 : 0000000000000001 a4 : ffffffda75b8f908
 a5 : ffffffff8247e908 a6 : 00000000065c7fe4 a7 : 0000000000000383
 s2 : ffffffd9039767a0 s3 : ffffffd9039767a0 s4 : 0000000000000001
 s5 : 0000000000000001 s6 : ffffffd90511ca00 s7 : 0000000000000002
 s8 : ffffffd90511cb20 s9 : 0000000000000005 s10: 0000000000000003
 s11: 0000000000000001 t3 : ffffffda75b8f908 t4 : 0000000000000000
 t5 : ffffffd905e7d668 t6 : ffffffd90511cdfc
status: 0000000200000120 badaddr: 0000000000000000 cause: 0000000000000003
[<ffffffff80eecbd8>] reg_process_self_managed_hint+0x154/0x1aa
[<ffffffff80eecd0e>] regulatory_set_wiphy_regd_sync+0x2c/0x8e
[<ffffffff03a428ca>] rtw_update_wiphy_regd+0x2c6/0x53c [8852bs]
[<ffffffff03a42b86>] rtw_regd_change_complete_sync+0x46/0x120 [8852bs]
[<ffffffff03a41338>] rtw_wiphy_register+0x10e/0x118 [8852bs]
[<ffffffff03a41860>] rtw_cfg80211_dev_res_register+0x18/0x24 [8852bs]
[<ffffffff03a2e3ca>] rtw_os_ndevs_register+0x1c/0x118 [8852bs]
[<ffffffff03a2e52a>] rtw_os_ndevs_init+0x30/0x44 [8852bs]
[<ffffffff03a4a108>] rtw_dev_probe+0x108/0x3e4 [8852bs]
[<ffffffff80bded30>] sdio_bus_probe+0xf4/0x110
[<ffffffff80872afc>] really_probe+0x8c/0x326
[<ffffffff80872df8>] __driver_probe_device+0x62/0x110
[<ffffffff80872ed2>] driver_probe_device+0x2c/0xb2
[<ffffffff80872fca>] __device_attach_driver+0x72/0xd0
[<ffffffff80870ce6>] bus_for_each_drv+0x5c/0xb2
[<ffffffff808733a2>] __device_attach+0x86/0x15a
[<ffffffff8087361e>] device_initial_probe+0xe/0x16
[<ffffffff80871b0c>] bus_probe_device+0x88/0x8a
[<ffffffff8086f392>] device_add+0x536/0x6f6
[<ffffffff80bdeeae>] sdio_add_func+0x62/0x7c
[<ffffffff80bddf4e>] mmc_attach_sdio+0x182/0x350
[<ffffffff80bd3a72>] mmc_rescan+0x274/0x334
[<ffffffff800397aa>] process_one_work+0x1b0/0x45e
[<ffffffff80039bf6>] worker_thread+0x19e/0x350
[<ffffffff80041190>] kthread+0xda/0xf6
[<ffffffff80fb66e6>] ret_from_fork+0xe/0x1c

Change-Id: Ie9cecd4946839a1e05b22706f7c4f9786664de0d
---
 drivers/net/wireless/realtek/rtl8852bs/os_dep/linux/wifi_regd.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/net/wireless/realtek/rtl8852bs/os_dep/linux/wifi_regd.c b/drivers/net/wireless/realtek/rtl8852bs/os_dep/linux/wifi_regd.c
index 95eeca0f1cf8..ab358e130372 100644
--- a/drivers/net/wireless/realtek/rtl8852bs/os_dep/linux/wifi_regd.c
+++ b/drivers/net/wireless/realtek/rtl8852bs/os_dep/linux/wifi_regd.c
@@ -393,7 +393,9 @@ void rtw_update_wiphy_regd(struct wiphy *wiphy, struct get_chplan_resp *chplan,
 		rtnl_lock();
 
 	#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 12, 0))
+	wiphy_lock(wiphy);
 	ret = regulatory_set_wiphy_regd_sync(wiphy, regd);
+	wiphy_unlock(wiphy);
 	#else
 	ret = regulatory_set_wiphy_regd_sync_rtnl(wiphy, regd);
 	#endif
-- 
2.47.0

